polygeist/tools/cgeist/Test/Verification/ptrarith.c

```diff
@@ -0,0 +1,76 @@
+// RUN: cgeist %s --function=* -S | FileCheck %s
+
+#include <stddef.h>
+
+// CHECK-LABEL:   func.func @f0(
+// CHECK-SAME:                  %[[VAL_0:.*]]: memref<?xi32>,
+// CHECK-SAME:                  %[[VAL_1:.*]]: i32) -> memref<?xi32>
+// CHECK:           %[[INDEX:.*]] = arith.index_cast %[[VAL_1]] : i32 to index
+// CHECK:           %[[ADD:.*]] = "polygeist.subindex"(%[[VAL_0]], %[[INDEX]]) : (memref<?xi32>, index) -> memref<?xi32>
+// CHECK:           return %[[ADD]] : memref<?xi32>
+// CHECK:         }
+
+int *f0(int *ptr, int index) {
+  return ptr + index;
+}
+
+// CHECK-LABEL:   func.func @f1(
+// CHECK-SAME:                  %[[VAL_0:.*]]: i64,
+// CHECK-SAME:                  %[[VAL_1:.*]]: memref<?xi32>) -> memref<?xi32>
+// CHECK:           %[[INDEX:.*]] = arith.index_castui %[[VAL_0]] : i64 to index
+// CHECK:           %[[ADD:.*]] = "polygeist.subindex"(%[[VAL_1]], %[[INDEX]]) : (memref<?xi32>, index) -> memref<?xi32>
+// CHECK:           return %[[ADD]] : memref<?xi32>
+// CHECK:         }
+
+int *f1(size_t index, int *ptr) {
+  return index + ptr;
+}
+
+// CHECK-LABEL:   func.func @f2(
+// CHECK-SAME:                  %[[VAL_0:.*]]: i64) -> !llvm.ptr<i8>
+// CHECK:           %[[PTR_0:.*]] = llvm.inttoptr %[[VAL_0]] : i64 to !llvm.ptr<i8>
+// CHECK:           return %[[PTR_0]] : !llvm.ptr<i8>
```

Comment:
Did `GEP` on `null` (incorrect)

Reply:
We're lowering `void *` to `!llvm.ptr<i8>`. I'd argue this makes sense because (what I understand reading some documentation-discussions in MLIR threads) `memref` implies there is an underlying object (it is a **reference**, not a **pointer**). 

Anyway, what code we generate for `void *` is outside of the scope of this PR.